### PR TITLE
Add internal calendar sharing

### PR DESCRIPTION
**Pull request by bertrandda**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to [the community](https://community.gladysassistant.com/) for testing before merging.
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.js`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Added attribute in model `Calendars` to list Gladys users who can see the calendar.
In front calendar page display my calendars & calendars shared with me

image

### CODE REVIEW

**Comment by Pierre-Gilles** [comment]
First, thanks for the PR ! 🙏

I don't really agree with the database modelization

First question, couldn't we just add a boolean like "shared: true/false", that just share the calendar with the whole instance? 

Do we really need this level of granularity?

**Comment by bertrandda** [comment]
When it's a classic home usage maybe it's not necessary, but I though about someone with a deep usage of Gladys (ex: Terdious here https://community.gladysassistant.com/t/multi-utilisateurs/5510/5). Maybe he will want to share one calendar only with his wife and an other with his children or colleagues. Granularity was easy to set up in calendar so why not ?

I agree on modelisation, the easiest way was to just add an attribute that concatenate ids to not create specific joining table but I can create one if you prefer.

**Comment by Pierre-Gilles** [comment]
I'm not sure we want to go for the most complex solution to solve the issue for all users, I'm sure most people just need a shared: true/false. 

In the case someone needs a very specific sharing, maybe he can just handle the sharing part in his own calendar (send invites), and then let each user connect the same calendar to Gladys (and in that case, we need to handle duplicate external_id)

I see two solutions: 

- Either we stick to the user calendar sharing in Gladys, but in that case I'm more in favor of a `t_calendar_user` table (the relational way, with enforced foreign keys)
- Either we simplify and move to a boolean option (shared: true/false)

What do you think ?

**Comment by Pierre-Gilles** [comment]
I also see a potential problem: If you create a scene that works with a calendar (no the case right now, but it'll happen soon), if the user doesn't have access to the calendar, he won't be able to see the scene in the UI (and scene are cross-users)

If we create a notion of "public" calendars, we can potentially say that only public calendars can be used in scene

**Comment by bertrandda** [comment]
And we do not plan to have "private" scenes?

If no, indeed public true/false calendar could be the best way

**Comment by Pierre-Gilles** [comment]
No I don't think so.

Either the user is "admin" and he can see all scenes in the instance, or he is "user" and he is not authorized to create scenes.

**Comment by bertrandda** [comment]
Ok then boolean should be enough, I change that

File: front/src/config/i18n/en.json [hunk 1]
```diff
@@ -737,9 +737,11 @@
       "title": "CalDAV",
       "accountTab": "Account",
       "syncTab": "Synchronization",
+      "shareTab": "Sharing",
       "description": "Sync your CalDAV calendars.",
       "accountIntroduction": "To sync Gladys calendar using CalDAV, you first need an external compatible calendar. Then enter the configuration below.",
       "syncIntroduction": "Choose the calendars to synchronize.",
+      "shareIntroduction": "Choose the calendars you want to share with other Gladys users.",
```
Comment (atrovato) [inline front/src/config/i18n/en.json]: `Choose calendars` - no `the`

File: front/src/actions/calendar.js [hunk 2]
```diff
@@ -9,9 +9,12 @@ function createActions(store) {
       try {
         const events = await state.httpClient.get('/api/v1/calendar/event', {
           from,
-          to
+          to,
+          shared: true
+        });
+        const calendars = await state.httpClient.get('/api/v1/calendar', {
+          shared: true
```
Comment (atrovato) [inline front/src/actions/calendar.js:16]: Are you sure we need to explicitly request for shared calendars?
Shouldn't it the default behavior? "get mines AND shared"?

**Review by atrovato — COMMENTED** [review]
Great code, some comments.
I don't try it for now, just review code.
I will ASAP.

File: front/src/actions/calendar.js (hunk 2 above)
Comment (bertrandda) [reply front/src/actions/calendar.js:16]: In my opinion, default should be yours only. For example in configuration we fetch only your calendars to select those to sync and It.
Discussion is open, I can change it

Comment (Pierre-Gilles) [reply front/src/actions/calendar.js:16]: I agree with @bertrandda, what I see: 

- By default on the configuration page, you should only be able to change your calendars (not the shared one)
- On the calendar page, all calendars (yours + shared) are displayed
- In scenes, only shared calendars are displayed (not in this PR)

**Comment by bertrandda** [comment]
Not stale, ready for reviews

**Review by Pierre-Gilles — APPROVED** [review]
I just tested the PR end-to-end locally, and it works great! 👏

Amazing work, sorry for taking so long to review !

**Pull request merged**